README.md
# python-rate-limiter

Python rate-limiter with bucketing.


CHANGES.txt
# Revision history for python-rate-limiter

## 0.1.0.0 -- YYYY-mm-dd

* First version


tests/limiter_test.py
import rate_limiter as rl


def test_rate_limiter():
    endpoint = "/test"
    app = "app"
    r = rl.RateLimiter(
        id_or_name="test_limiter",
        endpoint_name_filter=endpoint,
        app_name_filter=app,
        user_filter=True,
        limit=10,
        window="minute",
        window_buckets=10
    )
    state = rl.ProcessState()
    req = rl.Request(app_id=app, endpoint=endpoint, user_id="1")
    rl.is_limit_reached(state, [r], req)


rate_limiter/service.py
"""
Public API for Rate Limits
--------------------------
"""
import logging
import time
from typing import Sequence

from .definitions import Request
from .definitions import RateLimiter
from .definitions import IState
from .definitions import Timestamp
from .definitions import Window
from .definitions import WindowRange

log = logging.getLogger(__name__)


def is_limit_reached(
    state: IState,
    rate_limits: Sequence[RateLimiter],
    request: Request
) -> bool:
    """ Checks if a given request is made outside existing rate limits.

    :param state: Rate Limiters' global shared state.
    :param rate_limits: sequence of currently active rate limiters.
    :param request: A request description object necessary for rate limiter functionality.
    """
    # Since we are using a generator expression to perform "check & hit",
    # we will return from this function on the first non-passed rate limiter.
    # Therefore, we need to make sure that we hit the smallest rate limit
    # first, to have a deterministic behaviour with multiple rate limits
    # that may / may not register the hit based on the success of the hit
    # of the preceding rate limiter. Hence, this sorting:
    rate_limits = sorted(rate_limits, key=lambda x: x.limit)

    now = time.time()
    # keep it a lazy object here, no lists
    limits_pass = (
        _hit(state, now, rl, request) for rl in rate_limits
        if _match(rl, request)
    )
    return not all(limits_pass)


def _match(rate_limiter: RateLimiter, request: Request) -> bool:
    """ Decides if the current rate limit is applicable to a given request.
    """
    if (
        rate_limiter.endpoint_name_filter
        and rate_limiter.endpoint_name_filter != request.endpoint
    ):
        return False

    if (
        rate_limiter.app_name_filter
        and rate_limiter.app_name_filter != request.app_id
    ):
        return False

    return True


def _hit(state: IState, now: float, rl: RateLimiter, request: Request) -> bool:
    """ Register an endpoint hit and tell if it is beyond allowed limit.

    :return: True if the hit was made within allowed limit, False otherwise.
    """
    try:
        window_range = _get_window_range(now, rl)
        domain_key = _get_counter_domain_key(rl, request)

        # Calculate the hit
        # -----------------
        counters = state.hit_and_collect_counters(domain_key, window_range)

        total_hits = 0
        expired = []
        for bucket, hits in counters.items():
            if Timestamp(bucket) < window_range.first_bucket:
                expired.append(bucket)
            else:
                total_hits += int(hits)

        # Invalidate (if necessary)
        # -------------------------
        state.expire_buckets(domain_key, expired)

        return total_hits <= rl.limit
    except Exception as e:
        log.error(str(e))
    return True


def _get_window_range(now: float, rl: RateLimiter) -> WindowRange:
    # we infer the number of seconds that fit into a single bucket of a provided window size,
    # note that different windows have different number of possible bucket sizes, and that affects
    # the number of seconds we can fit into one bucket
    seconds_per_bucket = _SECONDS_IN_WINDOW_SIZE[rl.window] // rl.window_buckets
    # once we have a number of seconds per bucket per window size, we can infer the current id
    # of the bucket by chunking the entire timeline into buckets
    # (because we know a typical bucket size in terms of seconds)
    current_serial_bucket_pos = int(now // seconds_per_bucket)

    return WindowRange(
        current_bucket=current_serial_bucket_pos * seconds_per_bucket,
        first_bucket=(current_serial_bucket_pos - rl.window_buckets + 1) * seconds_per_bucket,  # noqa
        lifetime=rl.window_buckets * seconds_per_bucket,
    )


def _get_counter_domain_key(rl: RateLimiter, request: Request) -> str:
    """ Counter Domain Key is the key of a hashkey object
    that is used for storing bucket-based counters for a given rate limiter.
    """
    endpoint_name = rl.endpoint_name_filter or _NOT_LIMITED_BY_ENDPOINT_NAME
    app_name = rl.app_name_filter or _NOT_LIMITED_BY_APP_NAME
    if rl.user_filter:
        user_id = request.user_id
    else:
        user_id = _NOT_LIMITED_BY_USER
    # We use {rl.id_or_name} prefix to prevent situations like:
    # 1. a rate limit with 100/hour is created on endpoint A
    # 2. a rate limit with 1000/day is created on endpoint A
    # 3. sum(counters) on the endpoint A is now mixed
    #    with 100/hour and 1000/day metrics.
    # We also don't want to use just {rl.id_or_name}, because modifications
    # made to endpoint name / app name / user_id will be mixed
    # within one hash object, which would provide the same mixed counters
    # issue across multiple subsequent modifications of the settings.
    return f'LIMIT:{rl.id_or_name}:{app_name}:{endpoint_name}:{user_id}'


# A few constants for underlying Redis storage implementation.
# Tilde is used for preventing id_or_name clash with actual endpoint / app id_or_name etc
_NOT_LIMITED_BY_ENDPOINT_NAME = '~anyendpoint'
_NOT_LIMITED_BY_APP_NAME = '~anyapp'
_NOT_LIMITED_BY_USER = '~anyuser'

_SECONDS_IN_WINDOW_SIZE = {
    Window.SECOND: 1,
    Window.MINUTE: 60,
    Window.HOUR: 60 * 60,
    Window.DAY: 60 * 60 * 24,
    Window.MONTH: 60 * 60 * 24 * 30,
    Window.YEAR: 60 * 60 * 24 * 365,
}


rate_limiter/in_redis.py
from typing import Mapping, Sequence

import redis

from .definitions import IState, WindowRange


class RedisState(IState):
    """ State implementation for Redis backing store.
    """
    __slots__ = ['redis_pool']

    def __init__(self) -> None:
        self.redis_pool = redis.Redis(connection_pool=redis.ConnectionPool())

    def hit_and_collect_counters(
        self,
        domain_key: str,
        window_range: WindowRange
    ) -> Mapping[str, int]:
        with self.redis_pool.pipeline(transaction=True) as pipe:
            pipe.hincrby(domain_key, str(window_range.current_bucket), 1)
            pipe.expire(domain_key, window_range.lifetime)
            pipe.hgetall(domain_key)
            *__, counters = pipe.execute()
            return counters  # type: ignore

    def expire_buckets(self, domain_key: str, buckets: Sequence[str]) -> None:
        if buckets:
            self.redis_pool.hdel(domain_key, *buckets)


rate_limiter/in_memory.py
import threading
from collections import defaultdict
from typing import Mapping, Sequence

from .definitions import IState, WindowRange


class ProcessState(IState):
    """ State implementation for process' in-memory store.
    """
    __slots__ = ['state', 'lock']

    def __init__(self):
        self.state: dict[str, dict[str, int]] = {}
        self.lock = threading.Lock()

    def hit_and_collect_counters(
        self,
        domain_key: str,
        window_range: WindowRange
    ) -> Mapping[str, int]:
        with self.lock:
            domain_buckets = self.state.setdefault(domain_key, defaultdict(int))
            domain_buckets[str(window_range.current_bucket)] += 1
            return domain_buckets

    def expire_buckets(self, domain_key: str, buckets: Sequence[str]) -> None:
        domain_buckets = self.state.get(domain_key)
        if not domain_buckets:
            return

        with self.lock:
            for x in buckets:
                try:
                    domain_buckets.pop(x)
                except KeyError:
                    pass


rate_limiter/definitions.py
from collections import defaultdict

import redis

from enum import Enum
import threading
from typing import NamedTuple, Sequence, Mapping, Protocol
from typing import Optional

Timestamp = int
Seconds = int


class WindowRange(NamedTuple):
    current_bucket: Timestamp
    first_bucket: Timestamp
    lifetime: Seconds


class Window(Enum):
    """ Available window types
    """
    SECOND = 'second'
    MINUTE = 'minute'
    HOUR = 'hour'
    DAY = 'day'
    MONTH = 'month'
    YEAR = 'year'


class RateLimiter:
    WINDOW_BUCKETS_LIMITS = {
        Window.SECOND: 60,
        Window.MINUTE: 60,
        Window.HOUR: 60,
        Window.DAY: 24,
        Window.MONTH: 30,
        Window.YEAR: 12,
    }
    __slots__ = (
        'id_or_name',
        'endpoint_name_filter',
        'app_name_filter',
        'user_filter',
        'limit',
        'window',
        'window_buckets',
    )

    def __init__(
        self,
        *,
        id_or_name: str,
        endpoint_name_filter: Optional[str],
        app_name_filter: Optional[str],
        user_filter: bool,
        limit: int,
        window: str,
        window_buckets: int,
    ) -> None:
        self.id_or_name = id_or_name
        self.endpoint_name_filter = endpoint_name_filter
        self.app_name_filter = app_name_filter
        self.user_filter = user_filter
        self.limit = limit
        self.window = Window(window)  # rate = `limit` per `window`
        if self.window is Window.SECOND:
            # we don't need several expiration buckets for a second
            window_buckets = 1
        self.window_buckets = window_buckets
        if window_buckets > self.WINDOW_BUCKETS_LIMITS[self.window]:
            raise ValueError(
                f'Unsupported buckets number {window_buckets} for window type {self.window}'
            )

    def __repr__(self) -> str:
        return (
            f'RateLimiter({self.id_or_name}, {self.endpoint_name_filter}, '
            f'{self.app_name_filter}, '
            f'{self.user_filter}, '
            f'{self.limit}, '
            f'{self.window.value})'
        )


class Request(NamedTuple):
    app_id: str
    """ Application ID of the client making a request.
    """
    user_id: str
    """ ID of a client making a request.
    """
    endpoint: str
    """ An identifier of the endpoint being rate-limited. It could be either a route name or route path, there's no
    big difference as long as it provides a unique name among a collection of all endpoints being rate-limited.
    """


class IState(Protocol):
    """ Generic global shared state interface for rate limiter data.
    """
    def hit_and_collect_counters(
        self,
        domain_key: str,
        window_range: WindowRange
    ) -> Mapping[str, int]:
        """ Increments the current window's current bucket and returns all existing buckets of the
        rate limiter registered under ``domain_key``
        """
        ...

    def expire_buckets(self, domain_key: str, buckets: Sequence[str]) -> None:
        """ Expire buckets that are out of range of the current window.
        """
        ...
